From 7a3225b3c242a368b3e08f04a0e3efd754d96142 Mon Sep 17 00:00:00 2001
From: Mao Minkai <maominkai@wxiat.com>
Date: Tue, 7 Jun 2022 17:42:43 +0800
Subject: [PATCH] sw64: vdso: fix backtrace of vrt_sigreturn
Patch-mainline: NO, OTHERS
References: openEuler-22.03-LTS
Modified-by-SEL: No


Sunway inclusion
category: bugfix
bugzilla: https://gitee.com/openeuler/kernel/issues/I56P0Z

--------------------------------

Add proper CFI directives to vdso rt_sigreturn to make backtrace work
properly.

Signed-off-by: Mao Minkai <maominkai@wxiat.com>

Signed-off-by: Gu Zitao <guzitao@wxiat.com>
Acked-by: Xie XiuQi <xiexiuqi@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 arch/sw_64/kernel/vdso/vrt_sigreturn.S | 32 ++++++++++++++++++++++++++
 1 file changed, 32 insertions(+)

diff --git a/arch/sw_64/kernel/vdso/vrt_sigreturn.S b/arch/sw_64/kernel/vdso/vrt_sigreturn.S
index c07eb7244d0c..90ec17e60e71 100644
--- a/arch/sw_64/kernel/vdso/vrt_sigreturn.S
+++ b/arch/sw_64/kernel/vdso/vrt_sigreturn.S
@@ -20,10 +20,42 @@
 #include <linux/linkage.h>
 #include <asm/unistd.h>
 
+#define RT_SIGFRAME_SIZE	1600
+#define RT_SIGFRAME_MCTX	176
+
 	.text
 
+	.macro SIGCONTEXT_REGS_I base, from = 0
+	.cfi_offset \from, \base + (4 + \from) * 8
+	.if	30 - \from
+	SIGCONTEXT_REGS_I \base, "(\from + 1)"
+	.endif
+	.endm
+
+	.macro SIGCONTEXT_REGS_F base, from = 32
+	.cfi_offset \from, \base + (4 + 1 + \from) * 32
+	.if	62 - \from
+	SIGCONTEXT_REGS_F \base, "(\from + 1)"
+	.endif
+	.endm
+
+	.macro SIGCONTEXT_REGS base
+	SIGCONTEXT_REGS_I \base
+	SIGCONTEXT_REGS_F \base
+	.cfi_offset 63, \base + (4 + 32 + 1) * 8 + 32 * 32
+	.cfi_offset 64, \base + 2 * 8
+	.endm
+
+	.cfi_startproc
+	.cfi_return_column 64
+	.cfi_signal_frame
+	SIGCONTEXT_REGS -RT_SIGFRAME_SIZE + RT_SIGFRAME_MCTX
+	.cfi_def_cfa_offset RT_SIGFRAME_SIZE
+
+	nop
 ENTRY(__vdso_rt_sigreturn)
 	mov		$sp, $16
 	ldi		$0, __NR_rt_sigreturn
 	sys_call	0x83
 ENDPROC(__vdso_rt_sigreturn)
+	.cfi_endproc
-- 
2.34.1

